pkg/controller/servicecatalog/provisionedproduct/utils.go

```diff
@@ -34,66 +36,50 @@ func provisioningParamsAreChanged(cfStackParams []cfsdkv2types.Parameter, curren
-		// access to both the product ID and all of the available provisioning
-		// artifact IDs so we will be able to discover both IDs with a single
-		// API call.
-		dpInput := &svcsdk.DescribeProductInput{}
-
-		if productID != "" {
-			dpInput.Id = aws.String(productID)
-		} else {
-			dpInput.Name = ds.ProductName
-		}
-
-		describeProduct, err := c.client.DescribeProduct(dpInput)
+func (c *custom) productOrArtifactAreChanged(ds *svcapitypes.ProvisionedProductParameters, resp *svcsdk.ProvisionedProductDetail) (bool, error) {
+	// ProvisioningArtifactID and ProvisioningArtifactName are mutual exclusive params, the same about ProductID and ProductName
+	// But if describe a provisioned product aws api will return only IDs, so it's impossible to compare names with ids
+	// Conditional statement below works only if desired state includes ProvisioningArtifactID and ProductID
+	if ds.ProvisioningArtifactID != nil && ds.ProductID != nil &&
+		(*ds.ProvisioningArtifactID != *resp.ProvisioningArtifactId ||
+			*ds.ProductID != *resp.ProductId) {
+		return true, nil
+		// In case if desired state includes not only IDs provider runs func `getArtifactID`, which produces
+		// additional request to aws api and retrieves an artifact id(even if it is already defined in the desired state)
+		// based on ProductId/ProductName for further comparison with artifact id in the current state
+	} else if ds.ProvisioningArtifactName != nil || ds.ProductName != nil {
+		desiredArtifactID, err := c.getArtifactID(ds)
 		if err != nil {
-			return true, errors.Wrap(err, "could not lookup product")
-		}
-		if describeProduct == nil || describeProduct.ProductViewSummary == nil {
-			return true, errors.New("could not find product")
+			return false, err
 		}
-
-		if productID == "" {
-			// fill in the product ID if we do not yet have it
-			if describeProduct == nil || describeProduct.ProductViewSummary == nil {
-				return true, errors.New("could not find product")
-			}
-			productID = pointer.StringDeref(describeProduct.ProductViewSummary.ProductId, "")
+		if desiredArtifactID != *resp.ProvisioningArtifactId {
+			return true, nil
 		}
+	}
+	return false, nil
+}
 
-		if artifactID == "" {
-			// find the matching artifact
-			for _, artifact := range describeProduct.ProvisioningArtifacts {
-				if pointer.StringEqual(ds.ProvisioningArtifactName, artifact.Name) {
-					artifactID = pointer.StringDeref(artifact.Id, "")
-					break
-				}
-			}
-		}
+func (c *custom) getArtifactID(ds *svcapitypes.ProvisionedProductParameters) (string, error) {
+	input := svcsdk.DescribeProductInput{
+		Id:   ds.ProductID,
+		Name: ds.ProductName,
+	}
+	// DescribeProvisioningArtifact methods fits much better, but it has a bug
+	output, err := c.client.DescribeProduct(&input)
+	if err != nil {
+		return "", errors.Wrap(err, errCouldNotLookupProduct)
+	}
+	if ds.ProvisioningArtifactName != nil && ds.ProvisioningArtifactID != nil {
```

Comment:
The best practice is to validate/sanitize the inputs as soon as possible, at the beginning, so this check should be done in the caller method immediately after the comment at line 40.